### PR TITLE
Reduce complexity of the 'scopes with var and let' challenge

### LINKED ISSUE #13025: [beta] Improve challengeSeed for "ES6: Scopes and the var keyword"
#### Issue Description
The editor shows an error for line 6 of the initial `challengeSeed`, stating that we shouldn't create functions within loops. We can avoid this by moving this function definition out of the loop and only reference it in the loop.

Updated `challengeSeed`:
```js
var newCampers = [{camper: "Wil"}, {camper: "Sam"}, {camper: "Dav"}];

function roleCall (i) {
  return "Camper # " + (i + 1) + " has arrived.";
} 

// only change code below this line
for (var i = 0; i < newCampers.length; i++) {
// only change code above this line
  newCampers[i].roleCall = roleCall.bind(null, i);
}

console.log(newCampers[0].roleCall());
console.log(newCampers[1].roleCall());
console.log(newCampers[2].roleCall());
```

However, this also means we lose the closure that gives us access to `i`. Since code isn't supposed to be written like this anyway, I think we're better of by using `roleCall.bind(null, i)`, despite the increased complexity. The code above works just like the previous example, but uses better (if not best) practices.

Even better, @BrendanSweeny suggested this could further be improved:
```js
var newCampers = [{camper: "Wil"}, {camper: "Sam"}, {camper: "Dav"}];

function roleCall (i) {
  return "Camper # " + (i + 1) + " has arrived.";
} 

// only change code below this line
for (var i = 0; i < newCampers.length; i++) {
// only change code above this line
  newCampers[i].roleCall = roleCall.bind(null, i);
}

console.log(newCampers[0].roleCall());
console.log(newCampers[1].roleCall());
console.log(newCampers[2].roleCall());
console.log(typeof i);
```

By adding a `console.log()`, the camper **can see** that using `let` will cause `typeof i` on the last line to be `"undefined"`.

@BrendanSweeny also had an idea about adding an extra test to verify that `i ===`. I think this is a good idea.

Building upon Brendan's test case from https://github.com/freeCodeCamp/freeCodeCamp/pull/12956#discussion_r98820528, this could be what we're looking for.
```js
assert.isUndefined(i, 'message: <code>i</code> should not be available outside of the <code>for</code> loop.');
```

### PR DESCRIPTION
**Pull request by BrendanSweeny**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #13025

#### Description

Rewrote the challengeSeed for the "Compare Scopes of the var and let Keywords" ES6 challenge to:
* Reduce the complexity of the challenge, which better highlights the different scopes assigned to the variables declared with `var` or `let`.
* Avoid the dubious coding practice of defining a function within a for loop and the associated linting error in the challenge's code.

Additional Changes:
* Expanded the "note" section to include a statement saying that in your own programs its probably best to use different names for the variables to avoid confusion.
* Added test case placeholders. These will need to be addressed in the future once transpiling issues are resolved.
* Very slight rewording of instructions.

### CODE REVIEW

File: seed/challenges/02-javascript-algorithms-and-data-structures/es6.json [hunk 1]
```diff
@@ -75,33 +75,32 @@
...
+        "// only change the code above this line",
+        "checkScope();"
       ],
       "tests": [
-        "assert(newCampers[0].roleCall() === \"Camper # 1 has arrived.\", 'message: <code>newCampers[0].call()</code> should call the index of the first camper');",
-        "assert(newCampers[1].roleCall() === \"Camper # 2 has arrived.\", 'message: <code>newCampers[1].call()</code> should call the index of the second camper');",
-        "assert(newCampers[2].roleCall() === \"Camper # 3 has arrived.\", 'message: <code>newCampers[2].call()</code> should call the index of the third camper');"
+        "// Test that <code>var</code> keyword is not used",
+        "// Test that <code>i</code> in if statement logs \"if statement scope\"",
+        "// Test that <code>i</code> in function logs \"function scope\""
```
Comment (Greenheart) [inline seed/challenges/02-javascript-algorithms-and-data-structures/es6.json]: `assert(code.match(/(i\\s*=\\s*)('|")function\\s*scope\\2;\\s*.*\\s*.*\\s*\\1/), 'message: The variable `i` declared in the function scope should equal `"function scope"`

We know that the string variable `code` always will contain all code read as if it's one long string. This means the declaration of the function-scoped `i` happens before the block-scoped `i`.

By using `code.match(/(i\s*=\s*)('|")function\s*scope\2;\s*.*\s*.*\s*\1/)` we can verify that the first of the `i`s has the right content.
function-scope

File: seed/challenges/02-javascript-algorithms-and-data-structures/es6.json [hunk 2]
```diff
@@ -75,33 +75,32 @@
...
-        "console.log(newCampers[2].roleCall());"
+        "// only change the code above this line",
+        "checkScope();"
       ],
       "tests": [
-        "assert(newCampers[0].roleCall() === \"Camper # 1 has arrived.\", 'message: <code>newCampers[0].call()</code> should call the index of the first camper');",
-        "assert(newCampers[1].roleCall() === \"Camper # 2 has arrived.\", 'message: <code>newCampers[1].call()</code> should call the index of the second camper');",
-        "assert(newCampers[2].roleCall() === \"Camper # 3 has arrived.\", 'message: <code>newCampers[2].call()</code> should call the index of the third camper');"
+        "// Test that <code>var</code> keyword is not used",
+        "// Test that <code>i</code> in if statement logs \"if statement scope\"",
```
Comment (Greenheart) [inline seed/challenges/02-javascript-algorithms-and-data-structures/es6.json]: `assert(code.match(/(i\\s*=\\s*).*\\s*.*\\s*.*\\1('|")if\\s*statement\\s*scope\\2/), 'message: The variable `i` declared in the function scope should equal `"if statement scope"`

Similarly, the assertion for the block scoped variable should be pretty similar except that we verify that it's the second `i`  variable, and that it has the `"if statement scope"`

block-scope

**Note:** Maybe we want to change this to the proper term - **block scope**?

**Review by Greenheart — CHANGES_REQUESTED** [review]
@BrendanSweeny Thanks - this is a great improvement! :smile:

I tested some regexes using an online tool, but I'm not sure that I escaped them for JSON properly so there might be issues when testing the assertions locally.

But we're getting closer! :+1:

File: seed/challenges/02-javascript-algorithms-and-data-structures/es6.json (hunk 1 above)
Comment (BrendanSweeny) [inline seed/challenges/02-javascript-algorithms-and-data-structures/es6.json]: I made a slight addition to the regex here. If we are testing that the `i` in the function scope does not change, we should specifically test for `let` in front of the `i` in the block scope. By doing this, the test will fail when `let` is omitted, which is the situation where function-scope-`i` will be reassigned to "block scope".

If you want to try it out in a regex tool, this is the expression: `/(i\\s*=\\s*)('|\")function\\s*scope\\2;\\s*.*\\s*.*\\s*let\\s+\\1/g`

I'll have the amended PR up shortly.

File: seed/challenges/02-javascript-algorithms-and-data-structures/es6.json (hunk 2 above)
Comment (BrendanSweeny) [inline seed/challenges/02-javascript-algorithms-and-data-structures/es6.json]: These regex expressions are great! Thanks for the help and the explanation. And I agree... I will rename 'if statement scope to 'block scope'. Also, the regex statements work fine without any additional escaping when testing locally.

**Comment by BrendanSweeny** [comment]
PR has been updated.

File: seed/challenges/02-javascript-algorithms-and-data-structures/es6.json (hunk 1 above)
Comment (Greenheart) [inline seed/challenges/02-javascript-algorithms-and-data-structures/es6.json]: @BrendanSweeny Great idea! I'll test all this locally :smile:

**Review by Greenheart — COMMENTED** [review]
@BrendanSweeny This looks good!

However, I'm not sure if we should merge this before we get the tests working, since it's now no longer possible to pass this challenge. What do you think?

**Comment by BrendanSweeny** [comment]
It depends. Yeah, testing for `let` obviously never passes at the moment, but do we have a guess how long it may take before we can run tests against non-transpiled code? Unless it's going to be soon, I vote we merge now and revisit the tests another time, since it sounds like providing new tests will be necessary for many of the challenges in this section anyway.

Alternatively, we could do something like this and negate the need to check for `let`:
```
function checkScope() {
  var i = "function scope";
  if (true) {
    let i = "block scope";
    console.log(i);
  }
  return i;
}

checkScope();

assert(!/var/g.test(code) && /let/g.test(code), 'message: The <code>var</code> keyword should be replaced with <code>let</code>.');
assert(code.match(/(i\\s*=\\s*).*\\s*.*\\s*.*\\1('|\")block\\s*scope\\2/g), 'message: The variable <code>i</code> declared in the if statement should remain equal to \"block scope\".');
assert(checkScope() === "function scope", 'message: <code>checkScope()</code> should return "function scope"');
```

If we have the function return one of the values for `i`, then we won't need to check for `let`. This makes the last test slightly less complex and the bottom two should work on the transpiled code (haven't tested locally though). I'm still slightly in favor of using two `console.log` statements because it makes the challenge more simple (albeit marginally) and the first test still isn't going to pass. Do you think this is a better alternative in the long run?

**Comment by Greenheart** [comment]
@BrendanSweeny I agree, we can always revisit this in the future.

I like that idea - that kind of test will allow use of `const`, while the first test ensures no `var`s have been used. Using return definitely makes a cleaner test case. 

@HKuz What do you think - should we use the existing changes in this PR or would @BrendanSweeny's latest suggestion be better in the long term?

**Comment by HKuz** [comment]
Hi guys, sorry I haven't been on top of this PR. Let me take a look - if you want to keep tests that will fail (because the transpilation issue isn't solved, and I'm not sure when that will be), but not actually have them run, you can always make that test a comment in the .json. So it would look like "//assert(...);" - you can add some text letting people know it's pending ("//TO COME assert(...);").

**Comment by Greenheart** [comment]
@HKuz Good idea!

What do you think about the latest proposed changes for the challenge's seed code?

**Comment by HKuz** [comment]
@Greenheart - I'll take a look at the two options and let you two know 👍

**Comment by HKuz** [comment]
Alright, gents - here's what I propose after comparing the suggestion above and the current PR. I think we can take the best of both worlds and combine them. Let me know if I'm missing something, or thoughts in general...

Let's keep the two `console.log` statements (like the current PR has) - if the output shows for the camper, that's a great demonstration of the scoping differences. We can make this really obvious by adjusting them slightly, so the block scope one is `console.log("Block scope i is:", i)` (should print "Block scope i is: block scope") and similar for the function scope one.

Then add the return statement at the end of the function, like the example above, and you can use the cleaner version of the last test.

Finally, I say we keep the `let/var` test since we'll eventually need and want it, but comment it out so people can pass the challenge. We can put an all-caps message in there so folks don't raise an issue, maybe "// TEMPORARILY COMMENTED assert(!var...);". We can use issue #12698 to keep track of bum tests until that transpilation issue is resolved.

**Comment by BrendanSweeny** [comment]
Thanks for taking a detailed look at it, @HKuz! I like your solution. I will go ahead and combine them as you suggest.

**Comment by Greenheart** [comment]
@HKuz Thanks for the help, that's an excellent solution!

I just added a line to your comment in #12698 about the disabled test.

**Review by Greenheart — APPROVED** [review]
LGTM!

@BrendanSweeny Thanks for sticking to it, despite all the iterations ^^

**Pull request merged**